File: c/roman-numerals/roman_numerals.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "roman_numerals.h"

char *roman[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
int arabic[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};

/*
 * I is 1
 * V is 5
 * X is 10
 * L is 50
 * C is 100
 * D is 500
 * M is 1000
 */
char *to_roman_numeral(unsigned int number) {
	char *roman_numeral = malloc(50);
	if (number == 0 || number > 3999) {
		roman_numeral = strcpy(roman_numeral, "");
		return roman_numeral; 
	}

	int num = number;
	for (int i=0; i<13; i++) {
		for (; num / arabic[i] > 0; num -= arabic[i]) {
			roman_numeral = strcat(roman_numeral, roman[i]);
		}
	}

	return roman_numeral;	
}

File: go/roman-numerals/roman_numerals.go
package romannumerals

import (
	"errors"
)

/*
 * I is 1
 * V is 5
 * X is 10
 * L is 50
 * C is 100
 * D is 500
 * M is 1000
 */
func ToRomanNumeral(number int) (string, error) {
	if number < 1 || number > 3000 {
		return "", errors.New("out of range")
	}
	var numeral string
	var i = number
	for ; (i / 1000) > 0; i -= 1000 {
		numeral += "M"
	}
	for ; (i / 900) > 0; i -= 900 {
		numeral += "CM"
	}
	for ; (i / 500) > 0; i -= 500 {
		numeral += "D"
	}
	for ; (i / 400) > 0; i -= 400 {
		numeral += "CD"
	}
	for ; (i / 100) > 0; i -= 100 {
		numeral += "C"
	}
	for ; (i / 90) > 0; i -= 90 {
		numeral += "XC"
	}
	for ; (i / 50) > 0; i -= 50 {
		numeral += "L"
	}
	for ; (i / 40) > 0; i -= 40 {
		numeral += "XL"
	}
	for ; (i / 10) > 0; i -= 10 {
		numeral += "X"
	}
	for ; (i / 9) > 0; i -= 9 {
		numeral += "IX"
	}
	for ; (i / 5) > 0; i -= 5 {
		numeral += "V"
	}
	for ; (i / 4) > 0; i -= 4 {
		numeral += "IV"
	}
	for ; (i / 1) > 0; i -= 1 {
		numeral += "I"
	}

	return numeral, nil
}
